File: my-solutions/0329-longest-increasing-path-in-a-matrix/solution.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        visited = set()
        max_path_rc = {}
        def dfs(r, c, prev_val):
            if r >= len(matrix) or c >= len(matrix[0]) or r < 0 or c < 0:
                return 0
            if prev_val >= matrix[r][c]:
                return 0
            if (r, c) in visited:
                return 0
            if (r, c) in max_path_rc:
                return max_path_rc[(r, c)]


            visited.add((r, c))
            res =  1 + max(
                dfs(r-1, c, matrix[r][c]),
                dfs(r+1, c, matrix[r][c]),
                dfs(r, c-1, matrix[r][c]),
                dfs(r, c+1, matrix[r][c])
            )
            visited.discard((r, c))
            max_path_rc[(r, c)] = res
            return res
        
        max_path = float("-inf")
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                visitied = set()
                max_path = max(max_path, dfs(i, j, -1))

        return max_path

File: my-solutions/0329-longest-increasing-path-in-a-matrix/solution.cpp
class Solution {
    int cache[201][201] = {0};
    int dirs[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        const int m = matrix.size(), n = matrix[0].size();

        int res = INT_MIN;
        for(int r = 0; r < m; ++r) {
            for(int c = 0; c < n; ++c) {
                res = max(res, dfs(r, c, matrix));
            }
        }
        return res;
    }
    int dfs(int r, int c, vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        if(cache[r][c]) return cache[r][c];

        int res = 1;
        for(int i = 0; i < 4; ++i) {
            int x = r + dirs[i][0], y = c + dirs[i][1];
            if(x >=0 && x < m && y >=0 && y < n && matrix[r][c] < matrix[x][y])
            res = max(res, 1 + dfs(x, y, matrix));
        }

        return cache[r][c] = res;
    }
};
